test/studies/ml/lib/Tensor.chpl

```diff
@@ -0,0 +1,792 @@
+        const (outH,outW): 2*int = correlateShape((kh,kw),(nh,nw),stride,padding);
+
+        var corr = new Tensor({0..#outH,0..#outW},real);
+
+        forall (x,y) in corr.data.domain with (ref corr) {
+            var sum = 0.0;
+            forall (i,j,c) in filter.data.domain with (+ reduce sum) {
+                sum += input[x * stride + i, y * stride + j,c] * filter[i,j,c];
+            }
+            corr[x,y] = sum;
+        }
+
+        return corr;
+    }
+
+    // Compute the resulting tensor shape of a kernel dialation
+    proc dialateShape(filterShape: 2*int, stride: int) {
+        const (kh,kw) = filterShape;
+        return (kh + (stride * (kh - 1)), kw + (stride * (kw - 1)));
+    }
+
+    // Dialate a filter
+    proc dialate(const ref filter: Tensor(2), stride: int = 1) {
+        const (kh,kw) = filter.shape;
+        const (dh,dw) = (kh + (stride * (kh - 1)), kw + (stride * (kw - 1)));
+        var d = new Tensor({0..#dh,0..#dw},real);
+        forall (i,j) in filter.data.domain with (ref d) {
+            d[i * stride, j * stride] = filter[i,j];
+            // d[i + i * stride,j + j * stride] = filter[i,j];
+        }
+        return d;
+    }
+
+    // Dialate a volume of filters
+    proc dialate(const ref filter: Tensor(3), stride: int = 1) {
+        const (kh,kw,kc) = filter.shape;
+        const (dh,dw) = (kh + (stride * (kh - 1)), kw + (stride * (kw - 1)));
+        var d = new Tensor({0..#dh,0..#dw,0..#kc},real);
+        forall (i,j,c) in filter.data.domain with (ref d) {
+            d[i * stride,j * stride,c] = filter[i,j,c];
+
+            // d[i + i * stride,j + j * stride,c] = filter[i,j,c];
+        }
+        return d;
+    }
+
+    // Compute the gradient of a loss with respect to a filter
+    proc filterGradient(const ref input: Tensor(2), const ref delta: Tensor(2), stride: int = 1, padding: int = 0) {
+        const d = dialate(delta,stride - 1);
+        return correlate(d,input,stride=1,padding=padding);
+    }
+
+    // Compute the gradient of a loss with respect to a volume of filters
+    proc filterGradient(const ref input: Tensor(3), const ref delta: Tensor(3), stride: int = 1, padding: int = 0,kernelSize: int) {
+        const (inH,inW,inC) = input.shape;
+        const (outH,outW,outC) = delta.shape;
+
+        const (dkh,dkw) = dialateShape((outH,outW),stride - 1);
+        const (kh,kw) = correlateShape((dkh,dkw),(inH,inW),stride=1,padding);
+
+        var grad = new Tensor(4,real);
+        if kh != kernelSize {
+            grad.reshapeDomain({0..#outC,0..#kernelSize,0..#kernelSize,0..#inC});
+            forall (ci,co) in {0..#inC,0..#outC} with (ref grad, var del = zeros(outH,outW), var img = zeros(inH,inW)) {
+                del = delta[..,..,co];
+                img = input[..,..,ci];
+                const d = dialate(del,stride - 1);
+                grad[co,..,..,ci] = correlate(d,img,stride=1,padding=padding)[0..#kernelSize,0..#kernelSize];
```

Comment:
I lost track again, but creating something only to grab a slice sounds like a potential for optimization. Are we creating redundant data only to throw it away?

Reply:
For these, I wasn't sure what was best. the `delta[..,..,co]` returns an array or an iterable expression, but `dialate` and `correlate` both require tensors as the parameter.

Reply:
Oh OK, I was thinking about the last line here strictly, but let's talk about what you're saying first.

`delta[..,..,co]` will return an "array view". These are essentially (record-wrapped) classes that don't have their own data, but have pointers to an existing data that belongs to an actual array, in this case, `delta`. However, when you assign an array view into a `var` (that's what you're doing here), it will actually copy the data pointed by the array view to create a new full-blown array. So, here, you are actually creating a new array in `del` and passing that to `dialate`. Is that what you really want? Or do you just want `dialate` to read a slice of the array, in which case, you are creating a redundant copy?

What I was saying here is `correlate(...)[0..#kernelSize, 0..#kernelSize];` looks like `correlate` returns an array, but then you slice a part of it (discarding the rest of the data) to assign it to part of `grad`. While this looks elegant, I can imagine a different `correlate` implementation that takes `grad`, `co`, `ci` as arguments, and computes the appropriate part of `grad` in-place without having to create a potentially larger array and then slicing it.

In any case, this part of the code can have big overheads due to extra array copies. However, rule no. 1 of performance optimizations is that you only optimize things if you know them to be an actual issue in application performance. IOW, the code here is elegant and unless the performance is an issue, no need to optimize.

Reply:
Ok, thank you for the clarification. I wasn't aware of the array view construct.